feat(validation): implement validate function and SchemaValidationErr…

VatsalGoel3 opened the pull request:
## Summary

### Changes

> Please provide a summary of what's being changed

<!-- What is this PR solving? Write a clear description or reference the issue(s) it addresses. -->

This PR improves the validate function by:

Ensuring correct handling of JMESPath expressions in the envelope parameter.
Adding type safety for formats to comply with Ajv's expectations.
Adding necessary type assertions for payload validation.
Fixing potential issues related to schema compilation failures.

Additionally, it adds comprehensive unit tests to cover:

Basic validation with valid and invalid payloads.
Validation with an envelope extracting nested objects.
Handling of custom formats and external schemas.
Proper error handling when schema compilation or validation fails.

> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** closes #3607

<!-------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/CONTRIBUTING.md#sending-a-pull-request
- Check that there isn't already a PR that addresses the same issue. If you find a duplicate, please leave a comment under the existing PR so we can discuss how to move forward
- Check that the change meets the project's tenets https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets
- Add a PR title that follows the conventional commit semantics - https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2
- If relevant, add tests that prove that the change is effective and works
- Whenever relevant, make sure to comment functions/methods/types and make appropriate changes to the documentation
------->

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

dreamorosi commented on packages/validation/src/index.ts:
```diff
@@ -1 +1,3 @@
 export const foo = () => true;
```
```suggestion
```

dreamorosi commented on packages/validation/src/validate.ts:
```diff
@@ -0,0 +1,55 @@
+  let validateFn: ValidateFunction;
+  try {
+    validateFn = ajvInstance.compile(schema);
+  } catch (error) {
```
Based on the coverage report I think we need to add a test case where we pass an invalid JSON Schema, which should make the `ajvInstance.compile(schema);` operation fail.

dreamorosi commented on packages/validation/src/validate.ts:
```diff
@@ -0,0 +1,55 @@
+import Ajv, { type ValidateFunction } from 'ajv';
+import { SchemaValidationError } from './SchemaValidationError';
+
+export interface ValidateParams<T = unknown> {
```
Can we extract this in a `types.ts` file in the same folder?

dreamorosi commented on packages/validation/src/validate.ts:
```diff
@@ -0,0 +1,55 @@
+import { search } from '@aws-lambda-powertools/jmespath';
+import Ajv, { type ValidateFunction } from 'ajv';
+import { SchemaValidationError } from './SchemaValidationError';
```
```suggestion
import { SchemaValidationError } from './errors';
```

dreamorosi commented on packages/validation/src/index.ts:
```diff
@@ -1 +1,3 @@
 export const foo = () => true;
+export * from './validate';
```
```suggestion
export { validate } from './validate.js';
```

dreamorosi commented on packages/validation/src/index.ts:
```diff
@@ -1 +1,3 @@
 export const foo = () => true;
+export * from './validate';
+export * from './SchemaValidationError';
```
```suggestion
export { SchemaValidationError } from './errors.ts';
```

dreamorosi commented on packages/validation/src/SchemaValidationError.ts:
Let's rename this file to `errors.ts` so we can put  more later if needed.

dreamorosi reviewed (changes requested):
Thank you so much for the PR @VatsalGoel3, I left a few comments but nothing major.

--

Regarding your question about tests, I think you might've been mistakenly running both unit and end to end tests and because of this, it was trying to deploy resources to an AWS account.

If you want to run the tests for this package, use this command: `npm run test:unit:coverage -w packages/validation` from the root of the project.

VatsalGoel3 commented:
I have updated the code with the required changes. Please let me know if any other changes are needed.

Thank you

dreamorosi commented on packages/validation/src/index.ts:
Please remove the first line - it was just a placeholder - and add `.js` to every import/export.

dreamorosi commented on packages/validation/src/validate.ts:
Please add `.js` to every local import and remove the `// Use default export` comment.

dreamorosi commented on packages/validation/src/validate.ts:
```diff
@@ -0,0 +1,53 @@
+  if (formats) {
+    for (const key of Object.keys(formats)) {
+      let formatDefinition = formats[key];
+      if (
```
Why did we add this check?

dreamorosi commented on packages/validation/tests/unit/validate.test.ts:
Please add `.js` to every local import.

dreamorosi reviewed (changes requested):
Thank you for addressing the comments, a couple more things that were not addressed in the previous iteration - I made them clear.

VatsalGoel3 commented:
Sorry for missing some changes; I updated them; also for the first line in index.ts, I had it because of the corresponding indes.test.ts; I have updated that test suite as well to match the updated index.ts

dreamorosi commented on packages/validation/src/index.ts:
```diff
@@ -1 +1,2 @@
-export const foo = () => true;
+export { validate } from './validate';
+export { SchemaValidationError } from './errors';
```
```suggestion
export { SchemaValidationError } from './errors.js';
```

dreamorosi commented on packages/validation/src/index.ts:
```diff
@@ -1 +1,2 @@
-export const foo = () => true;
+export { validate } from './validate';
```
```suggestion
export { validate } from './validate.js';
```

dreamorosi reviewed (approved):
Thank you for working on this PR!

If you're interested, you can take one between #3609 and #3608 which are a continuation of this one.
